src/sentry/api/endpoints/project_rule_actions.py

```diff
@@ -60,7 +65,58 @@ def post(self, request: Request, project) -> Response:
             project, platform=project.platform, default="javascript", tagged=True
         )
 
+        if features.has(
+            "projects:verbose-test-alert-reporting", project=project, user=request.user
+        ):
+            return self.execute_future_on_test_event(test_event, rule)
+        else:
+            # Old existing behavior to wrap this in a handler which buries exceptions
+            for callback, futures in activate_downstream_actions(rule, test_event).values():
+                safe_execute(callback, test_event, futures)
+            return Response()
+
+    def execute_future_on_test_event(
+        self,
+        test_event: GroupEvent,
+        rule: Rule,
+    ) -> Response:
+        """
+        A slightly modified version of utils.safe.safe_execute that handles
+        IntegrationFormErrors, and returns a body with `{ actions: [<error info>] }`.
+
+        This is used in our Alert Rule UI to display errors to the user.
+        """
+        action_exceptions = []
         for callback, futures in activate_downstream_actions(rule, test_event).values():
-            safe_execute(callback, test_event, futures)
+            try:
+                callback(test_event, futures)
+            except Exception as exc:
+                callback_name = getattr(callback, "__name__", str(callback))
+                cls_name = callback.__class__.__name__
+                logger = logging.getLogger(f"sentry.test_rule.{cls_name.lower()}")
+
+                # safe_execute logs these as exceptions, which can result in
+                # noisy sentry issues, so log with a warning instead.
+                if isinstance(exc, IntegrationError):
+                    logger.warning(
+                        "%s.test_alert.integration_error", callback_name, extra={"exc": exc}
```

Comment:
should we also check for `ApiError`? in code i often see us catch `ApiError` and subsequently raise `IntegrationError`, but i'm not sure if that's the case for every location `ApiError` could possible be raised for external API calls

Reply:
I'd rather restrict it to `IntegrationError` for now vs showing every API error our clients encounter. I'm already a bit wary of the amount of info we'll be propagating up to the UI with this 😅